alacritty/src/renderer/mod.rs

```diff
@@ -218,3 +220,48 @@ impl Renderer {
         }
     }
 }
+
+struct GlExtensions;
+
+impl GlExtensions {
+    /// Check if the given `extension` is supported.
+    ///
+    /// This function will lazyly load OpenGL extensions.
+    fn contains(extension: &str) -> bool {
+        static mut OPENGL_EXTENSIONS: Option<HashSet<&'static str>> = None;
+        static EXTENSIONS_LOADED: Once = Once::new();
+
+        EXTENSIONS_LOADED.call_once(|| unsafe {
+            let extensions = gl::GetString(gl::EXTENSIONS);
+
+            if extensions.is_null() {
+                let mut extensions_number = 0;
+                gl::GetIntegerv(gl::NUM_EXTENSIONS, &mut extensions_number);
+
+                OPENGL_EXTENSIONS = Some(
+                    (0..extensions_number as gl::types::GLuint)
+                        .flat_map(|i| {
+                            let extension =
+                                CStr::from_ptr(gl::GetStringi(gl::EXTENSIONS, i) as *mut _);
+                            extension.to_str()
+                        })
+                        .collect(),
+                );
+
+                return;
+            }
+
+            OPENGL_EXTENSIONS = match CStr::from_ptr(extensions as *mut _).to_str() {
+                Ok(ext) => Some(ext.split_whitespace().collect()),
+                Err(_) => Default::default(),
+            };
+        });
+
+        unsafe {
```

Comment:
Unfortunate we have to use unsafe here, but I don't think there's any way around this? There might be some crates that manage to hide that unsafe away, but with STD I don't recall a way. Did you check?

Reply:
Yeah, the safe way is to put into Mutex.

Reply:
I'd go with `OnceCell`, since it's a part of our Cargo.lock anyway and will go into rust stdlib eventually.

Reply:
That's what I was thinking of when talking about crates, thanks!